File: plugins/extract-django/catalog.py
"""The shapes of a fragment, in the order `catalog/model.go` declares them.

A mirror, not a second definition: `src/catalog.ts` is where the shape is
decided. Building the JSON here rather than at each call site is what keeps the
key order - and so the diff of a regenerated fragment - stable.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

DECLARED = "declared"
UNRESOLVED = "unresolved"


def field(name: str, type_: str, doc: str = "", ref: str = "") -> Dict[str, Any]:
    out = {"name": name, "type": type_, "doc": doc}
    if ref:
        out["ref"] = ref
    return out


def block(id_: str, slug: str, name: str, doc: str, fields: List[Dict[str, Any]]) -> Dict[str, Any]:
    return {"id": id_, "slug": slug, "name": name, "doc": doc, "fields": fields}


def event(id_: str, slug: str, name: str, versions: List[Dict[str, Any]], wire: Optional[Dict[str, str]]) -> Dict[str, Any]:
    out: Dict[str, Any] = {"id": id_, "slug": slug, "name": name, "versions": versions, "consumers": []}
    if wire:
        out["wire"] = wire
    return out


def version(doc: str, source: str, fields: List[Dict[str, Any]], name: str = "v1") -> Dict[str, Any]:
    return {"version": name, "doc": doc, "source": source, "fields": fields}


def operation(id_: str, kind: str, doc: str = "", exposed_by: Optional[List[str]] = None) -> Dict[str, Any]:
    out: Dict[str, Any] = {"id": id_, "kind": kind}
    if doc:
        out["doc"] = doc
    if exposed_by:
        out["exposedBy"] = sorted(exposed_by)
    return out


def aggregate(id_: str, slug: str, name: str, readme: str, root: str) -> Dict[str, Any]:
    return {
        "id": id_,
        "slug": slug,
        "name": name,
        "readme": readme,
        "root": root,
        "entities": [],
        "valueObjects": [],
        "operations": [],
        "events": [],
    }


def transition(from_: str, to: str, on: str, emits: str, source: str) -> Dict[str, Any]:
    out = {"from": from_, "to": to, "on": on}
    if emits:
        out["emits"] = emits
    if source:
        out["source"] = source
    return out


def rpc_call(id_: str, peer: str, status: str, source: str, note: str = "") -> Dict[str, Any]:
    out = {"id": id_, "peer": peer, "status": status, "source": source}
    if note:
        out["note"] = note
    return out


def participant(id_: str, kind: str, context: Optional[str], label: str = "") -> Dict[str, Any]:
    out: Dict[str, Any] = {"id": id_, "kind": kind, "context": context}
    if label:
        out["label"] = label
    return out


def step(id_: str, from_: str, to: str, kind: str, label: str, status: str, ref: str = "", note: str = "", line: str = "") -> Dict[str, Any]:
    out: Dict[str, Any] = {"type": "step", "id": id_, "from": from_, "to": to, "kind": kind}
    if label:
        out["label"] = label
    out["status"] = status
    if ref:
        out["ref"] = ref
    if note:
        out["note"] = note
    if line:
        out["line"] = line
    return out


def alt(id_: str, branches: List[Dict[str, Any]]) -> Dict[str, Any]:
    return {"type": "alt", "id": id_, "branches": branches}


def branch(title: str, steps: List[Dict[str, Any]], terminal: bool = False) -> Dict[str, Any]:
    out: Dict[str, Any] = {"title": title, "steps": steps}
    if terminal:
        out["terminal"] = True
    return out


def flow(id_: str, slug: str, name: str, summary: str, source: str, owner: str, participants: List[Dict[str, Any]], steps: List[Dict[str, Any]]) -> Dict[str, Any]:
    return {
        "id": id_,
        "slug": slug,
        "name": name,
        "summary": summary,
        "source": source,
        "owner": owner,
        "participants": participants,
        "steps": steps,
    }


def column(name: str, type_: str, nullable: bool, pk: bool = False, fk: Optional[Dict[str, str]] = None, maps: str = "", doc: str = "") -> Dict[str, Any]:
    out: Dict[str, Any] = {"name": name, "type": type_, "nullable": nullable}
    if pk:
        out["pk"] = True
    if fk:
        out["fk"] = fk
    if maps:
        out["maps"] = maps
    if doc:
        out["doc"] = doc
    return out


def table(id_: str, name: str, columns: List[Dict[str, Any]], indexes: List[Dict[str, Any]], persists: Optional[Dict[str, str]], role: str, doc: str = "") -> Dict[str, Any]:
    out: Dict[str, Any] = {"id": id_, "name": name}
    if doc:
        out["doc"] = doc
    out["columns"] = columns
    if indexes:
        out["indexes"] = indexes
    if persists:
        out["persists"] = persists
    if role:
        out["role"] = role
    return out


def store(id_: str, slug: str, name: str, kind: str, owner: str, tables: List[Dict[str, Any]], source: str) -> Dict[str, Any]:
    out = {"id": id_, "slug": slug, "name": name, "kind": kind, "owner": owner, "tables": tables}
    if source:
        out["source"] = source
    return out

File: plugins/extract-java/src/org/portolan/extract/Catalog.java
package org.portolan.extract;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The shapes of a fragment, in the order {@code catalog/model.go} declares them.
 *
 * A mirror, not a second definition: {@code src/catalog.ts} is where the shape
 * is decided. Building the JSON here rather than at each call site is what
 * keeps the key order - and so the diff of a regenerated fragment - stable.
 */
final class Catalog {

    static final String DECLARED = "declared";
    static final String UNRESOLVED = "unresolved";

    private Catalog() {}

    /** An ordered map from alternating keys and values. */
    static Map<String, Object> map(Object... pairs) {
        Map<String, Object> out = new LinkedHashMap<>();
        for (int i = 0; i + 1 < pairs.length; i += 2) {
            out.put(String.valueOf(pairs[i]), pairs[i + 1]);
        }
        return out;
    }

    static Map<String, Object> field(String name, String type, String doc) {
        return map("name", name, "type", type, "doc", doc);
    }

    static Map<String, Object> block(String id, String slug, String name, String doc, List<Object> fields) {
        return map("id", id, "slug", slug, "name", name, "doc", doc, "fields", fields);
    }

    static Map<String, Object> version(String doc, String source, List<Object> fields) {
        return map("version", "v1", "doc", doc, "source", source, "fields", fields);
    }

    static Map<String, Object> event(String id, String slug, String name, List<Object> versions, Map<String, Object> wire) {
        Map<String, Object> out = map("id", id, "slug", slug, "name", name, "versions", versions, "consumers", new ArrayList<>());
        if (wire != null) {
            out.put("wire", wire);
        }
        return out;
    }

    static Map<String, Object> operation(String id, String kind, String doc, List<String> exposedBy) {
        Map<String, Object> out = map("id", id, "kind", kind);
        if (!doc.isEmpty()) {
            out.put("doc", doc);
        }
        if (exposedBy != null && !exposedBy.isEmpty()) {
            out.put("exposedBy", exposedBy);
        }
        return out;
    }

    static Map<String, Object> aggregate(String id, String slug, String name, String readme, String root) {
        return map(
                "id", id,
                "slug", slug,
                "name", name,
                "readme", readme,
                "root", root,
                "entities", new ArrayList<>(),
                "valueObjects", new ArrayList<>(),
                "operations", new ArrayList<>(),
                "events", new ArrayList<>());
    }

    static Map<String, Object> transition(String from, String to, String on, String emits, String source) {
        Map<String, Object> out = map("from", from, "to", to, "on", on);
        if (!emits.isEmpty()) {
            out.put("emits", emits);
        }
        if (!source.isEmpty()) {
            out.put("source", source);
        }
        return out;
    }

    static Map<String, Object> rpcCall(String id, String peer, String status, String source) {
        return map("id", id, "peer", peer, "status", status, "source", source);
    }

    static Map<String, Object> participant(String id, String kind, String context, String label) {
        Map<String, Object> out = map("id", id, "kind", kind, "context", context);
        if (!label.isEmpty()) {
            out.put("label", label);
        }
        return out;
    }

    static Map<String, Object> step(String id, String from, String to, String kind, String label, String status, String ref, String note, String line) {
        Map<String, Object> out = map("type", "step", "id", id, "from", from, "to", to, "kind", kind);
        if (!label.isEmpty()) {
            out.put("label", label);
        }
        out.put("status", status);
        if (!ref.isEmpty()) {
            out.put("ref", ref);
        }
        if (!note.isEmpty()) {
            out.put("note", note);
        }
        if (!line.isEmpty()) {
            out.put("line", line);
        }
        return out;
    }

    static Map<String, Object> alt(String id, List<Object> branches) {
        return map("type", "alt", "id", id, "branches", branches);
    }

    static Map<String, Object> branch(String title, List<Object> steps, boolean terminal) {
        Map<String, Object> out = map("title", title, "steps", steps);
        if (terminal) {
            out.put("terminal", Boolean.TRUE);
        }
        return out;
    }

    static Map<String, Object> flow(String id, String slug, String name, String summary, String source, String owner, List<Object> participants, List<Object> steps) {
        return map(
                "id", id,
                "slug", slug,
                "name", name,
                "summary", summary,
                "source", source,
                "owner", owner,
                "participants", participants,
                "steps", steps);
    }
}
